### PR TITLE
Raise explicit error for pending message send fail

### PR DESCRIPTION
**Pull request by rhardih**
### Placeholder PR

If there's still pending messages after max retries have failed, raise
an explicit error instead of a generic RuntimeError.

### CODE REVIEW

File: lib/poseidon/sync_producer.rb [hunk 1]
```diff
@@ -63,7 +63,8 @@ def send_messages(messages)
       end
 
       if messages_to_send.pending_messages?
-        raise "Failed to send all messages: #{messages_to_send.messages} remaining"
+        msg = "Failed to send all messages: #{messages_to_send.messages} remaining"
+        raise Errors::MessagesSendError.new(msg)
```
Comment (bquorning) [inline lib/poseidon/sync_producer.rb]: `raise Errors::MessagesSendError, msg`

File: spec/unit/sync_producer_spec.rb [hunk 2]
```diff
@@ -58,6 +58,20 @@
         @sp = SyncProducer.new("test_client", [])
         @sp.send_messages([Message.new(:topic => "topic", :value => "value")]) rescue StandardError
       end
+
+      it "raises error if there's still pending messages" do
+        mbts = double('messages_to_send', :needs_metadata? => false).as_null_object
+
+        allow(mbts).to receive(:pending_messages?).and_return(true)
+        expect(mbts).to receive(:messages_for_brokers).once
```
Comment (bquorning) [inline spec/unit/sync_producer_spec.rb]: I believe you can use the existing `@mbts` instead:
```rb
allow(@mbts).to receive(:pending_messages?).and_return(true)
expect(@mbts).to receive(:messages_for_brokers).once
```

File: spec/unit/sync_producer_spec.rb [hunk 3]
```diff
@@ -58,6 +58,20 @@
...
+
+        allow(MessagesToSend).to receive(:new).and_return(mbts)
```
Comment (bquorning) [inline spec/unit/sync_producer_spec.rb]: This line’s also in the `before` block.

File: spec/unit/sync_producer_spec.rb (hunk 2 above)
Comment (bquorning) [inline spec/unit/sync_producer_spec.rb]: Actually, I think that expectation should just be a stubbing:
```rb
allow(@mbts).to receive(:messages_for_brokers).once
```
The real expectation is the `raise_error` one.

**Comment by bastien** [comment]
Looks good :+1:

**Comment by dasch** [comment]
Looks good!

File: spec/unit/sync_producer_spec.rb [hunk 4]
```diff
@@ -58,6 +58,16 @@
         @sp = SyncProducer.new("test_client", [])
         @sp.send_messages([Message.new(:topic => "topic", :value => "value")]) rescue StandardError
       end
+
+      it "raises error if there's still pending messages" do
+        allow(@mbts).to receive(:pending_messages?).and_return(true)
+        allow(@mbts).to receive(:messages_for_brokers) { [] }
```
Comment (bquorning) [inline spec/unit/sync_producer_spec.rb]: It looks like this projects uses `and_return` instead of the block-style returns.

**Comment by bquorning** [comment]
:+1: besides that one comment.

**Comment by rhardih** [comment]
Closing this and opening against upstream.

**Pull request closed without merging**